File: src/Code007_convert_queue_and_stack/Code01_ConvertQueueAndStack.java
package Code007_convert_queue_and_stack;

import java.util.Stack;

// 测试链接 : https://leetcode.cn/problems/implement-queue-using-stacks/

public class Code01_ConvertQueueAndStack {
    class MyQueue {

        public Stack<Integer> in;
        public Stack<Integer> out;

        public MyQueue() {
            in = new Stack<Integer>();
            out = new Stack<Integer>();
        }

        private void in2out() {
            if (out.empty()) {
                while (!in.empty())
                    out.push(in.pop());
            }
        }

        public void push(int x) {
            in.push(x);
            in2out();
        }

        public int pop() {
            in2out();
            return out.pop();
        }

        public int peek() {
            in2out();
            return out.peek();
        }

        public boolean empty() {
            return in.empty() && out.empty();
        }
    }
}

File: src/Code007_convert_queue_and_stack/Code01_ConvertQueueAndStack.cpp
#include <iostream>
#include <stack>
using namespace std;

class MyQueue
{
private:
    stack<int> in;
    stack<int> out;
    void In2Out()
    {
        if (out.empty())
        {
            while (!in.empty())
            {
                out.push(in.top());
                in.pop();
            }
        }
    }

public:
    MyQueue()
    {
    }

    void push(int x)
    {
        in.push(x);
        In2Out();
    }

    int pop()
    {
        In2Out();
        int x = out.top();
        out.pop();
        return x;
    }

    int peek()
    {
        In2Out();
        return out.top();
    }

    bool empty()
    {
        return in.empty() && out.empty();
    }
};